src/charm.py

```diff
@@ -979,6 +1057,118 @@ def auth_enabled(self) -> bool:
 
         return any("MONGOD_ARGS" in line and "--auth" in line for line in env_vars)
 
+    def _scope_obj(self, scope: Scopes):
+        if scope == APP_SCOPE:
+            return self.app
+        if scope == UNIT_SCOPE:
+            return self.unit
+
+    def _peer_data(self, scope: Scopes):
+        if not self._peers:
+            return {}.setdefault(scope, {})
+        scope_obj = self._scope_obj(scope)
+        return self._peers.data[scope_obj]
+
+    def _juju_secret_set(self, scope: Scopes, key: str, value: str) -> str:
+        """Helper function setting Juju secret."""
+        peer_data = self._peer_data(scope)
+        self._juju_secrets_get(scope)
+
+        secret = self.secrets[scope].get(Config.Secrets.SECRET_LABEL)
+
+        # It's not the first secret for the scope, we can re-use the existing one
+        # that was fetched in the previous call
+        if secret:
+            secret_cache = self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL]
+
+            if secret_cache.get(key) == value:
+                logging.debug(f"Key {scope}:{key} has this value defined already")
+            else:
+                secret_cache[key] = value
+                try:
+                    secret.set_content(secret_cache)
+                    logging.debug(f"Secret {scope}:{key} was {key} set")
+                except OSError as error:
+                    logging.error(
+                        f"Error in attempt to set '{key}' secret for scope '{scope}'. "
+                        f"Existing keys were: {list(secret_cache.keys())}. {error}"
+                    )
+
+        # We need to create a brand-new secret for this scope
+        else:
+            scope_obj = self._scope_obj(scope)
+
+            secret = scope_obj.add_secret({key: value})
+            if not secret:
+                raise SecretNotAddedError(f"Couldn't set secret {scope}:{key}")
+
+            self.secrets[scope][Config.Secrets.SECRET_LABEL] = secret
+            self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL] = {key: value}
+            logging.debug(f"Secret {scope}:{key} published (as first). ID: {secret.id}")
+            peer_data.update({Config.Secrets.SECRET_INTERNAL_LABEL: secret.id})
+
+        return self.secrets[scope][Config.Secrets.SECRET_LABEL].id
+
+    def _juju_secrets_get(self, scope: Scopes) -> Optional[bool]:
+        """Helper function to get Juju secret."""
```

Comment:
```suggestion
        """Helper function to retrieve all Juju secrets.
        
        Function returns a boolean based on X, but infact retrieves Juju secrets and stores them in the charm. (include expert knowledge)
        """
```

Reply:
Since this function has a lot of expert knowledge in it, i.e. the function makes has a specific design that relies on your expert knowledge, it would be nice to add a sentence or two to the docstring so that it is clear why the function works in this way

Reply:
To be fair - it is @juditnovak and @delgod knowledge, not mine 😄 

And it is copied from k8s charm as-is. I would prefer to keep them same for now. 
Return value does not make much sense and the function can be refactored, but not in this PR

Reply:
the function name can be better. in fact, it is just caching secret content into the internal dictionary.

Reply:
How about something like this?

```
This function is responsible for direct communication with the Juju Secret strore to retrieve the Mono Charm's single, unique Secret object's metadata, and --on success-- its contents.

In parallel with retrieving secret information, it's immediately locally cached, making sure that we have the snapshot of the secret for the lifetime of the event (that's being processed) without additional fetch requests to the Juju Secret Store.

(Note: metadata, i.e. the Secret object itself is cached as it may be necessary for later operations, like updating contents.)

The function is returning a boolean that reflects success or failure of the above.
```

This logic should be described somewhere else in a clear, conscize way. But as for now, as a firefighting it could provide "quick help" :-)

Reply:
I like Judits docstring here, I think it explains the complexities that are going on. Unless you feel strongly about removing it.